src/bsk201910/rollbacks.md

```diff
@@ -0,0 +1,407 @@
+
+The same property isn't available in the ReplicaSet.
+
+> How do you know which properties are available? [You can consult the official API](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#deployment-v1-apps).
+
+_In general, the YAML for the Deployment contains the ReplicaSet plus some additional details._
+
+You can create the ReplicaSet with:
+
+```terminal|command=1|title=bash
+kubectl create -f replicaset.yaml
+```
+
+_There's something else worth noting about the ReplicaSets and Deployments._
+
+When you upgrade your Pods from version 1 to version 2, the Deployment creates a new ReplicaSet and increases the count of replicas while the previous count goes to zero.
+
+**After the rolling update, the previous ReplicaSet is not deleted** — not immediately at least.
+
+Instead, it is kept around with a replicas count of 0.
+
+If you try to execute another rolling update from version 2 to version 3, you might notice that at the end of the upgrade, you have two ReplicaSets with a count of 0.
+
+_Why are the previous ReplicaSets not deleted or garbage collected?_
+
+Imagine that the current version of the container introduces a regression.
+
+You probably don't want to serve unhealthy responses to your users, so you might want to roll back to a previous version of your app.
+
+If you still have an old ReplicaSet, perhaps you could scale the current replicas to zero and increment the previous ReplicaSet count.
+
+In other words, **keeping the previous ReplicaSets around is a convenient mechanism to roll back to a previously working version of your app.**
+
+By default Kubernetes stores the last 10 ReplicaSets and lets you roll back to any of them.
+
+But you can change how many ReplicaSets should be retained by changing the `Deployment.spec.revisionHistoryLimit` in your deployment.
+
+```yaml|highlight=15|title=deployment.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: example-deployment
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      revisionHistoryLimit: 100
+      containers:
+      - name: app
+        image: learnk8s/hello:1.0.0
+```
+
+_What about the previous ReplicaSets?_
+
+_Could you list all the previous Replicasets that belong to a Deployment?_
+
+You can use the following command to inspect the history of your Deployment:
+
+```terminal|command=1|title=bash
+kubectl rollout history deployment/app
+```
+
+And you can rollback to a specific version with:
+
+```terminal|command=1|title=bash
+kubectl rollout undo deployment/app --to-revision=2
+```
+
+_But how does the Deployment know their ReplicaSets?_
+
+_Does it store the order in which ReplicaSets are created?_
+
+The ReplicaSets have random names with id such as `app-6ff88c4474`, so you should expect the Deployment to store a reference to them.
```

Comment:
Maybe you know this, but actually these names are not random: they are the hash of the pod template (only the pod template, no other parts).

Funny thing is: if you change the pod template back to something, that you had before, then the replicaset will be reused (because the name will match).  So you can do a "rollback" by chance, if you use e.g. the same image name again.

I'm not saying you have to put all these details into the article, but maybe replace "random" with "The ReplicaSet names are based on hashes of the pod template, such as..."

Reply:
The name goes through another step which is filtering for bad words such as "b00b5". So technically not always a hash.

Reply:
I'm not sure what to do here, as the intent is to have pseudo-random numbers. I'm planning to leave this as-in unless you feel strongly about having it as "The ReplicaSet names are based on hashes of the pod template, such as..."

Reply:
No, the intent is to have hashes.  Maybe we have the b00b5 implementation, I didn't know (btw do you have source code reference for this or any source, I would like to read up), but I just tested and if you change back to a previous template, you reuse the same replicaset.  So even if they have the boobs thing, they have it in the hash algorith, in a deterministic way.  So the intentional is not randomness, but hashing.

If you don't like "hash", but don't want "random", then maybe some handwaving: the "The ids in the ReplicaSet names (e.g. app-6ff...) are based on the pod template, but they don't point to the deployment, so you should expect the Deployment to store a reference to the them."

Actually I'm not sure what are you trying to say here and how the name hash/random is relevant to the Deployment having pointers or RSs having backward pointers, so I have no strong opinion either way.

If you want to resolve this quickly, then say "random looking" instead of "random".  Just "random" is quite bad, because makes you look lame, as they are not random. :)